Added n8n to samples

KevyVo opened the pull request:
By popular demand, [n8n](https://n8n.io/) is added to the list of samples. Since we do not support volumes right now, all the volumes in the compose.dev.yaml will be used for only local development. 


Here are some resources I used while developing this:

- https://github.com/n8n-io/n8n-hosting/tree/main/docker-compose/withPostgres
- https://docs.n8n.io/hosting/configuration/environment-variables/
- https://docs.n8n.io/hosting/installation/server-setups/docker-compose/
- https://community.n8n.io/t/error-while-deploying-n8n-docker/77926
- https://community.n8n.io/t/how-to-change-localhost-5678-from-webhook-url/27033/8

## Samples Checklist
✅ All good!

raphaeltm commented:
> I personally prefer if `compose.yaml` is for local/dev and `compose.prod.yaml` extends that one.

We had so many conversations about this and I used to push that idea, but we collectively agreed a long time ago that that was exactly what we wanted to avoid because it means passing an extra `-f` flag to get things deployed properly with Defang. That's why we structured every other sample with `compose.yaml` to cloud, and `compose.dev.yaml` for local dev.

KevyVo commented:
> Hey Kevin, Thanks for the mention. For the env vars in deploy.yaml, you can put them in the env section of the n8n service. For 1-click deploy, I'll help set that up once the PR is merged. Best, Raphaël Raphaël Titsworth-Morin (he/him) Co-Founder | Éphémère Creative Ltd [ephemerecreative.ca](https://ephemerecreative.ca) Let's build something good together.
> […](#)
> On Mon, Oct 6, 2025 at 10:25 AM ***@***.*** wrote: By popular require, n8n <https://n8n.io/> is added to the list of samples. Since we do not support volumes right now, all the volumes in the compose.dev.yaml will be used for only local development. Only two small items I help from @raphaeltm <https://github.com/raphaeltm> is to setup 1click deploy and set the env var in the deploy.yaml. I am unsure which environment to put it in. Here are some resources I used while developing this: - [https://github.com/n8n-io/n8n-hosting/tree/main/docker-compose/withPostgres](https://github.com/n8n-io/n8n-hosting/tree/main/docker-compose/withPostgres?rgh-link-date=2025-10-06T08%3A25%3A52Z) - https://docs.n8n.io/hosting/configuration/environment-variables/ - https://docs.n8n.io/hosting/installation/server-setups/docker-compose/ - https://community.n8n.io/t/error-while-deploying-n8n-docker/77926 - https://community.n8n.io/t/how-to-change-localhost-5678-from-webhook-url/27033/8 Samples Checklist ✅ All good! ------------------------------ You can view, comment on, or merge this pull request online at: #474 Commit Summary - [0f0ce0c](https://github.com/DefangLabs/samples/commit/0f0ce0cb115db6968bcfefae53e7fff50d1cfde1) <[0f0ce0c](https://github.com/DefangLabs/samples/pull/474/commits/0f0ce0cb115db6968bcfefae53e7fff50d1cfde1)> Added n8n to samples - [1695094](https://github.com/DefangLabs/samples/commit/1695094191f7fc03c88625ef7f4b9e1826b49f2a) <[1695094](https://github.com/DefangLabs/samples/pull/474/commits/1695094191f7fc03c88625ef7f4b9e1826b49f2a)> Working n8n with Oauth - [2d4a465](https://github.com/DefangLabs/samples/commit/2d4a465509fb7936f738dbf032d5d56ea9df4985) <[2d4a465](https://github.com/DefangLabs/samples/pull/474/commits/2d4a465509fb7936f738dbf032d5d56ea9df4985)> update reason for volumes - [b75f139](https://github.com/DefangLabs/samples/commit/b75f139b46b98e4ee744404719468ab22a24b4d5) <[b75f139](https://github.com/DefangLabs/samples/pull/474/commits/b75f139b46b98e4ee744404719468ab22a24b4d5)> Managed Postgres - [ffe0977](https://github.com/DefangLabs/samples/commit/ffe0977bf9ddf979432df784bcdc3b12efada7bd) <[ffe0977](https://github.com/DefangLabs/samples/pull/474/commits/ffe0977bf9ddf979432df784bcdc3b12efada7bd)> n8n Public DNS - [eb14aa0](https://github.com/DefangLabs/samples/commit/eb14aa06c26057c056ef991880862999613a7b73) <[eb14aa0](https://github.com/DefangLabs/samples/pull/474/commits/eb14aa06c26057c056ef991880862999613a7b73)> fix gcp setup deployment File Changes (6 files <https://github.com/DefangLabs/samples/pull/474/files>) - *A* samples/n8n/.github/workflows/deploy.yaml <https://github.com/DefangLabs/samples/pull/474/files#diff-3c5ddcabafecc6bf328a15063126c1c85f63143c97c67023d5fd425c83c27ea3> (28) - *A* samples/n8n/README.md <https://github.com/DefangLabs/samples/pull/474/files#diff-9468dc59215e010047102e5611169576d935d96c5c64d9d0dddaf9d8b3f09a33> (94) - *A* samples/n8n/compose.dev.yaml <https://github.com/DefangLabs/samples/pull/474/files#diff-b26b523e92e4d6b3f3c579cfd78fc2e974bbb63738241dc8be6a466c96f37bff> (40) - *A* samples/n8n/compose.yaml <https://github.com/DefangLabs/samples/pull/474/files#diff-c9e4bfd0cd96adb39bea69969645d51d6df017e1e9ccf4bf568eb42e3a2e40cd> (64) - *A* samples/n8n/init-data.sh <https://github.com/DefangLabs/samples/pull/474/files#diff-85c9cc6fa36ed5d05515650d7654d8bcb8554442fff1a29809245b4b77e1efea> (47) - *A* samples/n8n/setup.Dockerfile <https://github.com/DefangLabs/samples/pull/474/files#diff-80d0039114745e4d309f54ac715b7432358cfc71db63417028df03db274b15a0> (5) Patch Links: - https://github.com/DefangLabs/samples/pull/474.patch - https://github.com/DefangLabs/samples/pull/474.diff — Reply to this email directly, view it on GitHub <#474>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAG6MLZYLEDFUOSEPKJRW2D3WIRQFAVCNFSM6AAAAACHHLUEC6VHI2DSMVQWIX3LMV43ASLTON2WKOZTGQ2DGNZSGE4TCMQ> . You are receiving this because you were mentioned.Message ID: ***@***.***>

Are saying to hardcode the value to the `compose.yaml` to the n8n section env var like this:

```
n8n:
    image: docker.n8n.io/n8nio/n8n
    restart: always
    environment:
      - DB_TYPE=postgresdb
      - DB_POSTGRESDB_HOST=postgres
      - DB_POSTGRESDB_PORT=5432
      - DB_POSTGRESDB_DATABASE=n8n
      - DB_POSTGRESDB_USER=defang_user
      - DB_POSTGRESDB_PASSWORD=defang12345
      - N8N_HOST=n8n
      - N8N_PORT=5678
      - N8N_PROTOCOL=https
      - NODE_ENV=production
      - WEBHOOK_URL=https://n8n
      - DB_POSTGRESDB_SSL_ENABLED=true
      - DB_POSTGRESDB_SSL_REJECT_UNAUTHORIZED=false
    ports:
      - 5678:5678
    links:
      - postgres
    depends_on:
      setup:
        condition: service_completed_successfully
      postgres:
        condition: service_healthy
```

raphaeltm commented:
Hey Kevin,

Thanks for the PR! I've added the env vars in the deploy.yaml and will help
set up the 1-click deploy after merging. Looking forward to having n8n as
part of our samples.

Best,
Raphaël

Raphaël Titsworth-Morin (he/him)

Co-Founder | Éphémère Creative Ltd
https://ephemerecreative.ca
Let's build something good together.

On Mon, Oct 6, 2025 at 10:37 AM ***@***.*** wrote:

> *KevyVo* left a comment (DefangLabs/samples#474)
> <https://github.com/DefangLabs/samples/pull/474#issuecomment-3370503674>
>
> Hey Kevin, Thanks for the mention. For the env vars in deploy.yaml, you
> can put them in the env section of the n8n service. For 1-click deploy,
> I'll help set that up once the PR is merged. Best, Raphaël Raphaël
> Titsworth-Morin (he/him) Co-Founder | Éphémère Creative Ltd
> ephemerecreative.ca Let's build something good together.
> … <#>
> On Mon, Oct 6, 2025 at 10:25 AM *@*.*** wrote: By popular require, n8n
> https://n8n.io/ is added to the list of samples. Since we do not support
> volumes right now, all the volumes in the compose.dev.yaml will be used for
> only local development. Only two small items I help from @raphaeltm
> <https://github.com/raphaeltm> https://github.com/raphaeltm is to setup
> 1click deploy and set the env var in the deploy.yaml. I am unsure which
> environment to put it in. Here are some resources I used while developing
> this: -
> https://github.com/n8n-io/n8n-hosting/tree/main/docker-compose/withPostgres
> <https://github.com/n8n-io/n8n-hosting/tree/main/docker-compose/withPostgres?rgh-link-date=2025-10-06T08%3A25%3A52Z>
> - https://docs.n8n.io/hosting/configuration/environment-variables/ -
> https://docs.n8n.io/hosting/installation/server-setups/docker-compose/ -
> https://community.n8n.io/t/error-while-deploying-n8n-docker/77926 -
> https://community.n8n.io/t/how-to-change-localhost-5678-from-webhook-url/27033/8
> Samples Checklist ✅ All good! ------------------------------ You can view,
> comment on, or merge this pull request online at: #474
> <https://github.com/DefangLabs/samples/pull/474> Commit Summary - 0f0ce0c
> <https://github.com/DefangLabs/samples/commit/0f0ce0cb115db6968bcfefae53e7fff50d1cfde1>
> <0f0ce0c
> <https://github.com/DefangLabs/samples/pull/474/commits/0f0ce0cb115db6968bcfefae53e7fff50d1cfde1>>
> Added n8n to samples - 1695094
> <https://github.com/DefangLabs/samples/commit/1695094191f7fc03c88625ef7f4b9e1826b49f2a>
> <1695094
> <https://github.com/DefangLabs/samples/pull/474/commits/1695094191f7fc03c88625ef7f4b9e1826b49f2a>>
> Working n8n with Oauth - 2d4a465
> <https://github.com/DefangLabs/samples/commit/2d4a465509fb7936f738dbf032d5d56ea9df4985>
> <2d4a465
> <https://github.com/DefangLabs/samples/pull/474/commits/2d4a465509fb7936f738dbf032d5d56ea9df4985>>
> update reason for volumes - b75f139
> <https://github.com/DefangLabs/samples/commit/b75f139b46b98e4ee744404719468ab22a24b4d5>
> <b75f139
> <https://github.com/DefangLabs/samples/pull/474/commits/b75f139b46b98e4ee744404719468ab22a24b4d5>>
> Managed Postgres - ffe0977
> <https://github.com/DefangLabs/samples/commit/ffe0977bf9ddf979432df784bcdc3b12efada7bd>
> <ffe0977
> <https://github.com/DefangLabs/samples/pull/474/commits/ffe0977bf9ddf979432df784bcdc3b12efada7bd>>
> n8n Public DNS - eb14aa0
> <https://github.com/DefangLabs/samples/commit/eb14aa06c26057c056ef991880862999613a7b73>
> <eb14aa0
> <https://github.com/DefangLabs/samples/pull/474/commits/eb14aa06c26057c056ef991880862999613a7b73>>
> fix gcp setup deployment File Changes (6 files
> https://github.com/DefangLabs/samples/pull/474/files) - *A*
> samples/n8n/.github/workflows/deploy.yaml
> https://github.com/DefangLabs/samples/pull/474/files#diff-3c5ddcabafecc6bf328a15063126c1c85f63143c97c67023d5fd425c83c27ea3
> (28) - *A* samples/n8n/README.md
> https://github.com/DefangLabs/samples/pull/474/files#diff-9468dc59215e010047102e5611169576d935d96c5c64d9d0dddaf9d8b3f09a33
> (94) - *A* samples/n8n/compose.dev.yaml
> https://github.com/DefangLabs/samples/pull/474/files#diff-b26b523e92e4d6b3f3c579cfd78fc2e974bbb63738241dc8be6a466c96f37bff
> (40) - *A* samples/n8n/compose.yaml
> https://github.com/DefangLabs/samples/pull/474/files#diff-c9e4bfd0cd96adb39bea69969645d51d6df017e1e9ccf4bf568eb42e3a2e40cd
> (64) - *A* samples/n8n/init-data.sh
> https://github.com/DefangLabs/samples/pull/474/files#diff-85c9cc6fa36ed5d05515650d7654d8bcb8554442fff1a29809245b4b77e1efea
> (47) - *A* samples/n8n/setup.Dockerfile
> https://github.com/DefangLabs/samples/pull/474/files#diff-80d0039114745e4d309f54ac715b7432358cfc71db63417028df03db274b15a0
> (5) Patch Links: - https://github.com/DefangLabs/samples/pull/474.patch -
> https://github.com/DefangLabs/samples/pull/474.diff — Reply to this email
> directly, view it on GitHub <#474
> <https://github.com/DefangLabs/samples/pull/474>>, or unsubscribe
> https://github.com/notifications/unsubscribe-auth/AAG6MLZYLEDFUOSEPKJRW2D3WIRQFAVCNFSM6AAAAACHHLUEC6VHI2DSMVQWIX3LMV43ASLTON2WKOZTGQ2DGNZSGE4TCMQ
> . You are receiving this because you were mentioned.Message ID: *@*.***>
>
> Are saying to hardcode the value to the compose.yaml to the n8n section
> env var like this:
>
> n8n:
> image: docker.n8n.io/n8nio/n8n
> restart: always
> environment:
>   - DB_TYPE=postgresdb
>   - DB_POSTGRESDB_HOST=postgres
>   - DB_POSTGRESDB_PORT=5432
>   - DB_POSTGRESDB_DATABASE=n8n
>   - DB_POSTGRESDB_USER=defang_user
>   - DB_POSTGRESDB_PASSWORD=defang12345
>   - N8N_HOST=n8n
>   - N8N_PORT=5678
>   - N8N_PROTOCOL=https
>   - NODE_ENV=production
>   - WEBHOOK_URL=https://n8n
>   - DB_POSTGRESDB_SSL_ENABLED=true
>   - DB_POSTGRESDB_SSL_REJECT_UNAUTHORIZED=false
> ports:
>   - 5678:5678
> links:
>   - postgres
> depends_on:
>   setup:
>     condition: service_completed_successfully
>   postgres:
>     condition: service_healthy
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/DefangLabs/samples/pull/474#issuecomment-3370503674>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAG6ML47HBHS6QD577HCY2D3WIS4LAVCNFSM6AAAAACHHLUEC6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTGNZQGUYDGNRXGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

KevyVo commented:
Here are some issues I ran into over the last couple of days, along with how I resolved them:

	1.	Slow startup for managed Postgres
The managed Postgres instance takes a while to come online. I think it would be a better experience if we extend the retry period to up to 10 minutes and include a message in the script letting users know that the first initialization can take some time and to be patient. It’s much better than having the process fail and requiring another up. In my experience deploying this about 30 times over the past week, I’ve never had to wait more than 5 minutes on AWS or GCP.

	2.	GCP setup container environment variables
The GCP setup container kept failing because environment variables in the setup service aren’t populated unless we use the JavaScript-style ${} syntax. I had to inspect each Cloud Run service’s container environment to figure this out. I’m not sure if this is a bug or expected behavior — on AWS, we don’t need ${}, and the variables fill in fine. This only seems to affect GCP.

	3.	Init script re-deployment handling
I updated the init script to handle cases after the initial deployment. Once Postgres has been set up and the users are created, rerunning the script causes Postgres to error out since the users already exist. I added logic to handle that gracefully.

KevyVo commented:
Remaining issues:

1.	AWS redeploy issue
The initial deployment on AWS works perfectly with no problems. However, when running defang up a second time to redeploy, the n8n app crashes after deployment. A full defang down followed by defang up is required for it to deploy successfully again. I’ve checked the logs but haven’t yet determined the cause.

2.	GCP delayed startup
On GCP, the deployment completes successfully without any errors. However, once it’s finished, the deployed URL doesn’t load immediately. Interestingly, the app keeps trying to restart itself, and after about 30 minutes, it finally comes online and remains stable. I currently have a GCP n8n instance that’s been running for two days without issues. I’m still not sure why there’s such a long delay initially.
